actors/builtin/miner/miner_actor.go

```diff
@@ -1097,315 +1470,331 @@ func (a Actor) OnDeferredCronEvent(rt Runtime, payload *CronEventPayload) *adt.E
-			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to charge fault fee")
-			return nil
-		})
+	// Return pledge.
+	notifyPledgeChanged(rt, pledge.Neg())
 
-		terminateSectors(rt, expiredFaults, power.SectorTerminationFaulty, epochReward, pwrTotal.QualityAdjPower)
-		burnFundsAndNotifyPledgeChange(rt, ongoingFaultPenalty)
+	// Terminate deals.
+	for _, params := range dealsToTerminate {
+		requestTerminateDeals(rt, params.Epoch, params.DealIDs)
 	}
 
-	{
-		// Establish new proving sets and clear proofs.
-		rt.State().Transaction(&st, func() interface{} {
-			deadlines, err := st.LoadDeadlines(store)
-			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load deadlines")
+	// reschedule cron worker, if necessary.
+	return more
+}
 
-			// Assign new sectors to deadlines.
-			newSectors, err := st.NewSectors.All(NewSectorsPerPeriodMax)
-			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to expand new sectors")
+// Invoked at the end of the last epoch for each proving deadline.
+func handleProvingDeadline(rt Runtime) {
+	currEpoch := rt.CurrEpoch()
+	store := adt.AsStore(rt)
 
-			if len(newSectors) > 0 {
-				//if period end is not in the future, get randomness from previous epoch
-				randomnessEpoch := minEpoch(deadline.PeriodEnd(), rt.CurrEpoch()-ElectionLookback)
-				assignmentSeed := rt.GetRandomness(crypto.DomainSeparationTag_WindowedPoStDeadlineAssignment, randomnessEpoch, nil)
-				err = AssignNewSectors(deadlines, info.WindowPoStPartitionSectors, newSectors, assignmentSeed)
-				builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to assign new sectors to deadlines")
+	epochReward := requestCurrentEpochBlockReward(rt)
+	pwrTotal := requestCurrentTotalPower(rt)
+
+	hadEarlyTerminations := false
+
+	powerDelta := PowerPair{big.Zero(), big.Zero()}
+	newlyVested := big.Zero()
+	penaltyTotal := abi.NewTokenAmount(0)
+	pledgeDelta := abi.NewTokenAmount(0)
 
-				// Store updated deadline state.
-				err = st.SaveDeadlines(store, deadlines)
-				builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to store new deadlines")
+	var st State
+	rt.State().Transaction(&st, func() interface{} {
+		var err error
+		{
+			// Vest locked funds.
+			// This happens first so that any subsequent penalties are taken from locked pledge, rather than free funds.
+			newlyVested, err = st.UnlockVestedFunds(store, rt.CurrEpoch())
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to vest funds")
+		}
+
+		// Record whether or not we _had_ early terminations in the queue before this method.
+		// That way, don't re-schedule a cron callback if one is already scheduled.
+		hadEarlyTerminations = havePendingEarlyTerminations(rt, &st)
+
+		// Note: because the cron actor is not invoked on epochs with empty tipsets, the current epoch is not necessarily
```

Comment:
What if we're multiple deadlines ahead? I'm guessing we should skip the missed deadlines and give the miner a pass, but we should probably do something.